Make sure interned constants are immutable

RalfJung opened the pull request:
This makes sure that interning for constants (not statics) creates only immutable allocations.

Previously, the "main" allocation of `const FOO: Cell<i32> = Cell::new(0);` was marked as mutable, but I don't think we want that. It can be only copied, not written to.

Also, "leftover" allocations (behind raw pointers etc) were left mutable. I don't think we want to support that. I tried asserting that these are all already immutable (to double-check our static checks), but that failed in this one:
```rust
const NON_NULL_PTR2: NonNull<u8> = unsafe { mem::transmute(&0) };
```
Seems like maybe we want more precise mutability annotation inside Miri for locals (like `&0` here) so that this would actually become immutable to begin with?

I also factored `intern_shallow` out of the visitor so that we don't have to construct a visitor when we do not plan to visit anything. That confused me at first.

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

RalfJung commented:
r? @oli-obk

RalfJung commented on src/librustc_mir/const_eval.rs:
```diff
@@ -655,7 +653,7 @@ pub fn const_eval_raw_provider<'tcx>(
 
     let res = ecx.load_mir(cid.instance.def, cid.promoted);
     res.and_then(
-        |body| eval_body_using_ecx(&mut ecx, cid, body, key.param_env)
```
This was redundant, as `ecx.param_env` is already `key.param_env`. So I removed the parameter here and propagated that to the interner.

RalfJung commented on src/librustc_mir/interpret/intern.rs:
```diff
@@ -224,69 +262,44 @@ for
         None => (Mutability::Immutable, InternMode::ConstBase),
         // `static mut` doesn't care about interior mutability, it's mutable anyway
         Some(hir::Mutability::MutMutable) => (Mutability::Mutable, InternMode::Static),
     };
```
Does this correctly compute `ConstBase` for things like array sizes and non-`Copy` array initializers? What does it do for promoteds?

oli-obk replied:
for promoteds we also run into the `None` branch, as the `DefId` is of the function of the promoted, which doesn't have `static_mutability`.

The only way we cannot get `ConstBase` is by the `DefId` pointing to a static. So not sure what happens to `non-Copy` array initializers, as that's promotion happening inside a static

RalfJung replied:
> the DefId is of the function of the promoted, which 

But what if we are promoting inside the body of a `static`?

> not sure what happens to non-Copy array initializers, as that's promotion happening inside a static

@eddyb do you know more here?

oli-obk replied:
Do we actually create implicit promoteds inside statics? I don't think so. I think we just nuke a bunch of storage statements

RalfJung replied:
Hm... seems like that is the case, and also for `const`:
```rust
const fn id<T>(x: T) -> T { x }

const FOO: &'static usize = {
    let v = id(&std::mem::size_of::<i32>());
    v
};
```
[Check the MIR](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e5f4c87bd7b22e1d144f885703f90b87)

But... @eddyb isn't that in contradiction to what you told me about the "enclosing scope" stuff?

eddyb replied:
Maybe I should open a PR about fixing that logic, or an issue at least.
We should only be nuking `StorageDead` for locals that we want to promote the *borrow* of, nothing else.

I guess if I want `[None; 123]` to work for any element type in a constant, we'd have to do *actual* promotion, hmpf, we can't do nothing (since then MIR borrowck will still error like it does today) nor remove the `StorageDead`'s (since that has no effect either).

eddyb replied:
@RalfJung Sorry, everyone is replying and GitHub is not updating, so assume most of my comments are minutes older than GitHub says.

Your use of `id` is entirely unnecessary, introducing an inner `let` drops the `'static` scope anyway, so `&std::mem::size_of::<i32>()` has to be promoted (by removing `StorageDead`).

Note that something like `const X: &T = &expr;` doesn't have `StorageDead` on the local being borrowed *from MIR building*, whereas anything more nested requires promotion *to reach that* (but both use the lack of a `StorageDead` to signal that).

RalfJung replied:
> has to be promoted (by removing StorageDead).

Wait, what? I thought "being promoted" means firing up the promotion machinery, where parts of the local MIR are moved into a new body and computed separately and so on. That's what promotion *is*, we even have [a type for that](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/struct.Promoted.html) these days and [a query](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.promoted_mir). Are you saying that use of the term "promoted" is wrong -- or at least, that not all "promoteds"  actually have a corresponding `Promoted`? That would be really bad use of terminology...

It seems that even are lots of discussion, we are still not using the term "promotion" the same way. I'd really appreciate if you could define precisely what you mean by that term, because this is getting frustrating.^^

RalfJung replied:
Alternatively I propose we use "promoted" to refer to what is called a `Promoted` in rustc, and then we probably have to call whatever happens in my example above something else. "Lifetime extension" or so.

eddyb replied:
Promotion in consts and statics *is performed* by removing the `StorageDead` - that's what the bitset returned by `mir_const_qualif` is for - since that already has to work.

Maybe it shouldn't, especially since it's not enough for the `[None; 123]` case I mentioned earlier.

It would actually be less code overall to be uniform (but it might be slightly more expensive). I can prepare a PR if you want.

RalfJung replied:
> Promotion in consts and statics is performed by removing the StorageDead - that's what the bitset returned by mir_const_qualif is for - since that already has to work.

So I was using the term for the mechanism, and you were using for the effect. That would explain the confusion.
Nuking `StorageDead` actually makes sense to me, and that's what I used to think happens in consts -- but then I talked with you and you taught me about the "enclosing scope" rule and since then I was sure it was *not* done with `StorageDead`.

I think "promotion" makes more sense as the name for the mechanism of extracting part of the MIR out -- and that's what `Promoted` already means in the compiler. So I feel like we should come up with another name for the effect of something having a longer lifetime than expected. But we can also do it the other way around, as long as we are able to clearly distinguish the two.

We are looking for three terms:
* The mechanism that extracts parts of a MIR and moves it out (the kind of thing `promoted_mir` gives access to, from what I understand).
* The specific case of using that mechanism on unsuspecting run-time code, which has the effect of moving it to compile-time, leading to a very restrictive analysis (this is what `promoted.md` in this repo talks about).
* The analysis that determines which things get `'static` lifetime even though they usually wouldn't. The result of this analysis is used to nuke `StorageDead` in `const` (and `static`?), and to extract part of a MIR and cut it out elsewhere.

Finally I feel like we are making progress on this minefield of three different things that are all referred to by the same name.^^

RalfJung replied:
So, coming back to the [original question](https://github.com/rust-lang/rust/pull/63955#discussion_r318256605): does this code correctly result in `ConstBase` for all the right things? It seems like the answer is yes right now, but how fragile is that? Does `static_mutability` guarantee to return `None` for "MIR promoteds"?

RalfJung replied:
> I guess if I want [None; 123] to work for any element type in a constant, we'd have to do actual promotion, hmpf, we can't do nothing (since then MIR borrowck will still error like it does today) nor remove the StorageDead's (since that has no effect either).

So currently non-`Copy` array initializers don't work in statics/consts?

eddyb replied:
> So currently non-`Copy` array initializers don't work in statics/consts?

Correct, I don't see how they could.

RalfJung replied:
[Looks like you are right](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=cbb7b9a3a70fcdc0578109b425b9bbab).

Are there other cases we have to worry about here in terms of classifying CTFE-evaluated "stuff" into statics and consts ("everything but statics")?

oli-obk replied:
Kind of related is that consts can't refer to statics

RalfJung replied:
How is that related? Even if they would, that memory would already be interned.

oli-obk replied:
It was the only thing I could come up with where we decide only on explict static vs not

RalfJung replied:
"explict static vs not"?

oli-obk commented on src/librustc_mir/interpret/intern.rs:
```diff
@@ -45,58 +42,108 @@ enum InternMode {
+    // read-only memory, and also by Miri when evluating other constants/statics that
+    // access this one.
+    if mode == InternMode::Static {
+        let frozen = ty.map_or(true, |ty| ty.is_freeze(
```
leave a note saying that `None` is for vtables and thus we can assume frozenness (which is not 100% true, we may have created the vtable from a mutable static, but that would not end up here as it would already be interned)

RalfJung commented on src/test/ui/consts/miri_unleashed/mutable_const.rs:
```diff
@@ -0,0 +1,20 @@
+    unsafe {
+        *MUTABLE_BEHIND_RAW = 99 //~ WARN skipping const checks
+        //~^ ERROR any use of this value will cause an error
+        //~^^ tried to modify constant memory
```
I guess the interesting question this test exposes is: if there is a way to sneak by the "constants are immutable" check, the way this code here does, what is the desired behavior? Should it be UB to mutate that memory (that's what this PR does) or just observably different from when the constant got inlined everywhere (which used to be the case before)?

I wish we could just error when we find untyped pointers in a constant. But, well, this compiles on stable:
```rust
use std::ptr::NonNull;

union Transmute {
    reference: &'static u8,
    nonnull: NonNull<u8>,
}

const NON_NULL_PTR: NonNull<u8> = unsafe { Transmute { reference: &0 }.nonnull };
```
The allocation here probably arises from a `force_allocation`. I wonder if there are circumstances under which that could make the new allocation read-only? Probably not, before even for immutable bindings we often have to `force_allocation` before initializing, as we can only initialize the value in memory.

oli-obk replied:
There's no need for unsafe in order to get untyped pointers in Constants, just initializing a Union with a pointer and having a constant of Union type will get you that.

RalfJung replied:
Ah, good point:
```rust
union U {
    reference: &'static u8,
    int: usize,
}

const NON_NULL_PTR: U = U { reference: &0 };
```

oli-obk replied:
> Should it be UB to mutate that memory (that's what this PR does) or just observably different from when the constant got inlined everywhere (which used to be the case before)?

In my experience everyone has argued for that to be UB. In fact, once we unify promoteds between functions and constants to always create an anonymous immutable static, that will be the behaviour anyway

RalfJung replied:
Given that this was UB before your interning refactoring, I think I am fine making it UB again.

oli-obk commented on src/test/ui/consts/miri_unleashed/mutable_const.rs:
```diff
@@ -0,0 +1,20 @@
+
+// make sure we do not just intern this as mutable
+const MUTABLE_BEHIND_RAW: *mut i32 = &UnsafeCell::new(42) as *const _ as *mut _;
+
```
Huh, this compiles? I would have assumed const qualif to complain

RalfJung replied:
This is an "unleash Miri" test. ;)

eddyb replied:
This is under `miri_unleashed`, I'm assuming const-checking wouldn't error.

But also... if `miri_unleashed` *isn't* broken, then this should error as leaking the address of a local, no?

Unless the "outermost scope is `'static`" logic propagates through casts.

RalfJung replied:
> But also... if miri_unleashed isn't broken, then this should error as leaking the address of a local, no?

Not sure where that error should come from. The local is not destroyed before computation ends (there is no `StorageDead`).

eddyb replied:
But it does error if you use `id(&UnsafeCell::new(42))`, right? Then my last theory is correct (casts are transparent wrt to the scope thing).

oli-obk replied:
> This is an "unleash Miri" test. ;)

I knew that, I just assumed there'd be a warning about skipping static checks

RalfJung replied:
> I knew that, I just assumed there'd be a warning about skipping static checks

There is: `WARN skipping const checks`

oli-obk replied:
not on the `MUTABLE_BEHIND_RAW` constant

RalfJung replied:
> But it does error if you use id(&UnsafeCell::new(42)), right? Then my last theory is correct (casts are transparent wrt to the scope thing).

Then we get
```
error: any use of this value will cause an error
  --> src/test/ui/consts/miri_unleashed/mutable_const.rs:11:1
   |
11 | const MUTABLE_BEHIND_RAW: *mut i32 = id(&UnsafeCell::new(42)) as *const _ as *mut _;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered dangling pointer in final constant
   |
```

RalfJung replied:
> not on the MUTABLE_BEHIND_RAW constant

Yeah seems like we only get that on uses of constants, but that is consistent with your tests.

bors commented:
:umbrella: The latest upstream changes (presumably #63561) made this pull request unmergeable. Please resolve the merge conflicts.

Alexendoo commented:
Ping from triage, any updates? @RalfJung @oli-obk

RalfJung commented:
I'm on vacation, will rebase when I have time.

Other than that I think I addresses all comments, though [this question here](https://github.com/rust-lang/rust/pull/63955#discussion_r318958662) for @oli-obk and @eddyb is still open.

RalfJung commented:
Rebased, ready for review.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 5462ecb4b131a513b2d821d52b9af491781bd898 has been approved by `oli-obk`

<!-- @bors r=oli-obk 5462ecb4b131a513b2d821d52b9af491781bd898 -->
<!-- homu: {"type":"Approved","sha":"5462ecb4b131a513b2d821d52b9af491781bd898","approver":"oli-obk"} -->
